File: examples/typescript/case08.ts
export type RecoveryRecord = { state: string; revision: number };
export type RecoveryStore = { record: RecoveryRecord };

export class WeakRecoveryProjection {
  private guard = 0; // New process starts with no revision evidence.
  private readonly store: RecoveryStore;
  constructor(store: RecoveryStore) { this.store = store; }

  // faultscope:begin fs-c08.memory-only-revision
  apply(event: RecoveryRecord): boolean {
    if (event.revision <= this.guard) return false;
    this.store.record = { ...this.store.record, state: event.state }; // Revision is not saved.
    this.guard = event.revision;
    return true;
  }
  // faultscope:end fs-c08.memory-only-revision
}

// faultscope:begin fs-c08.persist-state-with-revision
export function persistStateWithRevision(store: RecoveryStore, next: RecoveryRecord): void {
  store.record = { ...next }; // One atomic durable record in this model.
}
// faultscope:end fs-c08.persist-state-with-revision

export class StrongRecoveryProjection {
  private ready = false;
  private current: RecoveryRecord = { state: "", revision: 0 };
  private readonly store: RecoveryStore;
  private constructor(store: RecoveryStore) { this.store = store; }

  // faultscope:begin fs-c08.recover-state-with-revision
  static recover(store: RecoveryStore): StrongRecoveryProjection {
    const projection = new StrongRecoveryProjection(store);
    projection.current = { ...store.record }; // Restore value and revision together.
    projection.ready = true; // Only then can the projection accept events.
    return projection;
  }
  // faultscope:end fs-c08.recover-state-with-revision

  // faultscope:begin fs-c08.reject-stale-after-restart
  apply(event: RecoveryRecord): boolean {
    if (!this.ready || event.revision <= this.current.revision) return false;
    persistStateWithRevision(this.store, event);
    this.current = { ...event };
    return true;
  }
  // faultscope:end fs-c08.reject-stale-after-restart
}

// faultscope:begin fs-c08.rebuild-before-ready
export function rebuildBeforeReady(history: RecoveryRecord[]): StrongRecoveryProjection {
  const store: RecoveryStore = { record: { state: "", revision: 0 } };
  for (const event of history) { // Complete authoritative history; no external effects here.
    if (event.revision > store.record.revision) persistStateWithRevision(store, event);
  }
  return StrongRecoveryProjection.recover(store); // Readiness follows replay.
}
// faultscope:end fs-c08.rebuild-before-ready

File: examples/python/case08.py
"""FS-C08: business-data survival does not imply revision-guard survival."""

from dataclasses import dataclass


@dataclass(frozen=True)
class RecoveryRecord:
    state: str
    revision: int


@dataclass
class RecoveryStore:
    record: RecoveryRecord


class WeakRecoveryProjection:
    def __init__(self, store: RecoveryStore) -> None:
        self.store = store
        self.guard = 0  # Each new process starts without revision evidence.

    # faultscope:begin fs-c08.memory-only-revision
    def apply(self, event: RecoveryRecord) -> bool:
        if event.revision <= self.guard:
            return False
        self.store.record = RecoveryRecord(event.state, self.store.record.revision)
        self.guard = event.revision  # Not durable.
        return True
    # faultscope:end fs-c08.memory-only-revision


# faultscope:begin fs-c08.persist-state-with-revision
def persist_state_with_revision(store: RecoveryStore, next_record: RecoveryRecord) -> None:
    store.record = next_record  # One atomic durable record in this model.
# faultscope:end fs-c08.persist-state-with-revision


class StrongRecoveryProjection:
    def __init__(self, store: RecoveryStore) -> None:
        self.store = store
        self.current = RecoveryRecord("", 0)
        self.ready = False

    # faultscope:begin fs-c08.recover-state-with-revision
    @classmethod
    def recover(cls, store: RecoveryStore) -> "StrongRecoveryProjection":
        projection = cls(store)
        projection.current = store.record  # Restore value and revision together.
        projection.ready = True  # Only after recovery may events be applied.
        return projection
    # faultscope:end fs-c08.recover-state-with-revision

    # faultscope:begin fs-c08.reject-stale-after-restart
    def apply(self, event: RecoveryRecord) -> bool:
        if not self.ready or event.revision <= self.current.revision:
            return False
        persist_state_with_revision(self.store, event)
        self.current = event
        return True
    # faultscope:end fs-c08.reject-stale-after-restart


# faultscope:begin fs-c08.rebuild-before-ready
def rebuild_before_ready(history: list[RecoveryRecord]) -> StrongRecoveryProjection:
    store = RecoveryStore(RecoveryRecord("", 0))
    for event in history:  # Complete authoritative history; no external effects here.
        if event.revision > store.record.revision:
            persist_state_with_revision(store, event)
    return StrongRecoveryProjection.recover(store)  # Ready only after replay.
# faultscope:end fs-c08.rebuild-before-ready
